sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventBatch.cs

```diff
@@ -157,6 +159,26 @@ public override bool TryAdd(EventData eventData)
                 _sizeBytes = size;
                 _batchMessages.Add(message);
 
+                // For a data body, the message converter will marshal the ReadOnlyMemory into an
+                // ArraySegment projection to avoid allocating.  To ensure the batch is immutable,
+                // it will need to own a buffer for the data body to guard against mutation of the
+                // source event.
+                //
+                // Because the batch is disposable and allows for deterministic clean-up, rent buffers
+                // from the shared pool rather than allocating a dedicated instance for each body segment.
+
+                if (eventData.GetRawAmqpMessage().Body.BodyType == AmqpMessageBodyType.Data)
+                {
+                   foreach (var bodySegment in message.DataBody)
+                   {
```

Comment:
Since the segments are known why not rent one buffer that is sliced for all of them?